js/csv.js

```diff
@@ -0,0 +1,230 @@
+    const pattern = new RegExp(
+        "(\\" +
+            delimiter +
+            "|\\r?\\n|\\r|^)" +
+            '(?:"([^"]*(?:""[^"]*)*)"|' +
+            '([^"\\' +
+            delimiter +
+            "\\r\\n]*))",
+        "gi"
+    );
+
+    // creates the initial data structure already populated
+    // with an initial value as we need to "look-back"
+    const data = [[]];
+
+    // creates an array to hold our individual pattern
+    // matching groups
+    let matches = null;
+
+    // keeps looping over the regular expression matches
+    // until we can no longer find a match
+    while ((matches = pattern.exec(dataS))) {
+        // retrieves the delimiter that was found, this will
+        // affect the newline operation
+        const matchedDelimiter = matches[1];
+
+        // checks to see if the given delimiter has a length
+        // (if it0s not the start of string) and if it matches
+        // field delimiter, if id does not, then we know
+        // that this delimiter is a row delimiter
+        if (matchedDelimiter.length && matchedDelimiter !== delimiter) {
+            // since we have reached a new row of data,
+            // adds an empty row to our data array
+            data.push([]);
+        }
+
+        // allocates space for the value that has been matched
+        let value;
+
+        // now that we have our delimiter out of the way,
+        // checks to see which kind of value we captured
+        // (can be quoted or unquoted)
+        if (matches[2]) {
+            // as a quoted value has been found we need to
+            // unescape the double quotes
+            value = matches[2].replace(/""/g, '"');
+        } else {
+            // a simple (non quoted) value is found and as
+            // such it should be considered the matched value
+            // without any kind of manipulation
+            value = matches[3];
+        }
+
+        // adds the newly (matched) value to the last item in
+        // the data array sequence
+        data[data.length - 1].push(value);
+    }
+
+    return object ? _toObject(data) : data;
+};
+
+/**
+ * Converts a sequence of parsed data items into a sequence of
+ * object like items structured according to the CSV.
+ *
+ * @param {Array} data The array of data items defined as sequences
+ * of data items, the first element should be the header.
+ * @returns {Array} An array of object items equivalent to the provided
+ * data array structure provided as input.
+ */
+export const _toObject = data => {
+    const objects = [];
+    const header = data[0];
+    const items = data.slice(1);
+    for (const item of items) {
+        const object = {};
+        for (let index = 0; index < header.length; index++) {
+            const key = header[index];
+            const value = item[index];
+            object[key] = value;
+        }
+        objects.push(object);
+    }
+    return objects;
+};
+
+export const _parseStringComma = (value, delimiter) => {

```

Comment:
Can't the delimiter be anything? Not just a comma? Maybe do `delimiter = ","` as a default value.

Also, what is the point of the function? To convert 3 into "3", "hello" into "hello" but "hello, world" into "hello world"?

Reply:
The point of this function is to not get incorrect values when analysing a string. For example, if there's a value called `hello, world`, in a CSV this will split `hello` and `world` into two different cells. What we want is to have that value in a single cell, so if we just convert it to `"hello, world"`, it will appear in one cell. If the value received is `3`, the value that will be received will still be `3`. 

This also respects any delimiter that might be used. If for example, we use a delimiter `;`, this function will only assist when that string is found inside a value, otherwise it will just return the value as is.